### ISSUE TITLE
NUTS initialization problem when sampling non-observed values from discrete distributions

### ISSUE THREAD

**Issue by vanAmsterdam** [pyro-ppl/numpyro#542]
In my model I need to sample (possibly observed) discrete variables. Using the `obs=y` where `y=None` possibly, works for Normal distributions, but not for Bernoulli (or Poisson).

This does work in Pyro by the way.

Here is a minimal example:

```
import jax.numpy as np
from jax import random, vmap
import numpyro
from numpyro.distributions import Normal, Bernoulli
from numpyro.infer.mcmc import MCMC, NUTS

rng_key = random.PRNGKey(seed=1)

# data
n     = 100
a     = 0.5
b     = 1.0
x     = Normal(0,1).sample(rng_key, sample_shape=(n,))
theta = a + b * x
y     = Bernoulli(logits=theta).sample(rng_key, sample_shape=(n,))

# models
def m1(x, y=None):
    a = numpyro.sample('a', Normal(0, 5))
    b = numpyro.sample('b', Normal(0, 5))
    with numpyro.plate('obs', n):
        theta = a + b * x
        numpyro.sample('y_obs', Normal(theta, 0.1), obs=y)

def m2(x, y=None):
    a = numpyro.sample('a', Normal(0, 5))
    b = numpyro.sample('b', Normal(0, 5))
    with numpyro.plate('obs', n):
        theta = a + b * x
        numpyro.sample('y_obs', Bernoulli(logits=theta), obs=y)

# sampling
kernel1 = NUTS(m1)
mcmc1   = MCMC(kernel1, num_warmup=10, num_samples=10)
mcmc1.run(rng_key, x,y) # works
mcmc1.run(rng_key, x)   # works


kernel2 = NUTS(m2)
mcmc2   = MCMC(kernel2, num_warmup=10, num_samples=10)
mcmc2.run(rng_key, x,y) # works
mcmc2.run(rng_key, x)   # biject_to(fn.support) not implemented, during funid_valid_initial_params
```

**Comment by fehiepsi** [comment]
@vanAmsterdam Currently, NumPyro does not support models with discrete latent variables (the case `y=None`). We are targeting this for the next release.

**Comment by neerajprad** [comment]
NUTS will not work with discrete variables. The reason why it works in Pyro is because Pyro marginalizes out all the discrete variables in the model automatically. We are looking to have support for this in NumPyro through integration with [funsor](https://github.com/pyro-ppl/funsor). This will provide a more general framework for discrete variable enumeration or marginalizing conjugate distributions, which @fehiepsi has been working on.

**Comment by vanAmsterdam** [comment]
ok thanks for the fast reply

the non-observed discrete variables are all leaf nodes in my model so I should be able move around this issue for now by returning the parameter that goes into the discrete distribution and then sample values outside of the MCMC sampling

btw this is probably a jax issue but looking at the numbers of my example it seems like something's off with jax.random.bernoulli (and thus numpyro.distributions.Bernoulli)

only the result from numpy makes any sense or am I missing something? (using jax 0.1.59 with jaxlib 0.1.38)

```
from jax import random
import numpy as onp
from jax.scipy.special import expit
from numpyro.distributions import Normal, BernoulliLogits

rng_key = random.PRNGKey(seed=1)
onp.random.seed(1)

# data
n     = 1000
a     = 1
b     = 0.05
x     = Normal().sample(rng_key, sample_shape=(n,))
theta = a + b * x
y1    = BernoulliLogits(theta).sample(rng_key)
y2    = random.bernoulli(rng_key, expit(theta))
y3    = onp.random.binomial(size=n, n=1, p=expit(theta))
print(x[y1==1].mean()) # -0.47936615
print(x[y1==0].mean()) # 1.210252
print(x[y2==1].mean()) # -0.47936615
print(x[y2==0].mean()) # 1.210252
print(x[y3==1].mean()) # -0.02344727
print(x[y3==0].mean()) # -0.11560267
```

**Comment by fehiepsi** [comment]
You are using the same random key for x and y1/y2, hence they are correlated. :)

**Comment by vanAmsterdam** [comment]
sorry should have been more clear. The point was that the means of `x[y==1]` and `x[y==0]` are very far apart, even though the should be very close according to the data generating mechanism. On top of this `x[y==1]` should be greater than `x[y==0]`. In fact, there is no overlap between the values of `x` between the groups of `y`:

```
print(x[y1==1].max())  # 0.63259774
print(x[y1==0].min())  # 0.64056104
```

I there something wrong with how I call the distributions or is there something weird going on?

**Comment by fehiepsi** [comment]
I think you will get the expected behavior if you replace
```
y1    = BernoulliLogits(theta).sample(rng_key)
```
by
```
y1    = BernoulliLogits(theta).sample(random.PRNGKey(2))
```
The situation is similar to when you create two uniform random variables, one in [0, 1] (`x = lambda: np.random.rand(1)`) and another one in [1, 2] (`y = lambda: np.random.rand(1) + 1`). If you use the same seed, you will always get `y() == x() + 1`.

**Comment by vanAmsterdam** [comment]
ok that works. So the random seed that makes `x = Normal()` be low or high makes `y = Bernoulli(x)` low or high at the same time if you reuse the same seed. Got it, thanks!

**Comment by stuhlmueller** [comment]
Just wanted to +1 this. Unobserved discrete latent variables are the most important feature I'd want before switching to Numpyro.

**Comment by fehiepsi** [comment]
Thanks for the example and feedback @vanAmsterdam! FYI, the issue has been resolved with #572. Until that PR is landed, let's move the discussion to #571 instead. :)
